web/src/composebox_typeahead.ts

```diff
@@ -1338,7 +1338,55 @@ export function initialize({
     };
     new Typeahead(stream_message_typeahead_input, {
         source(): string[] {
-            return topics_seen_for(compose_state.stream_id());
+            const topics = topics_seen_for(compose_state.stream_id());
+            const topic_input = stream_message_typeahead_input.$element.val()?.toString() ?? "";
+
+            let dm_option: string[] = [];
+            if (topic_input.length >= 3) {
+                const matched_users = people
+                    .get_realm_users()
+                    .filter((user) =>
+                        user.full_name.toLowerCase().startsWith(topic_input.toLowerCase()),
```

Comment:
You probably want to start by filtering using `build_person_matcher`, which is optimized for this use case in various ways (both correctness -- matching with diacritics in mind -- and performance).